File: src/hg/utils/automation/kmerPrint.pl
#!/usr/bin/env perl

use strict;
use warnings;

my $argc = scalar(@ARGV);

if ($argc < 2) {
  printf STDERR "usage: kmerPrint.pl <N> <fastaFile.fa> [moreFastaFiles.fa] | gzip -c > result.gz\n";
  printf STDERR "where <N> is kmer size\n";
  printf STDERR "followed by fasta file names (can be 'stdin')\n";
  exit 255;
}

# first argument is the id to place on each kmer
# first argument is kmer size
my $kmerSize = shift;

my $chrName = "";
my $chrStart = 0;

#######################################################################
sub countKmers($) {
  my ($sequence) = @_;
  my $sequenceLen = length($sequence);
  for (my $i = 0; $i <= ($sequenceLen - $kmerSize); ++$i) {
    my $kmer = substr($sequence, $i, $kmerSize);
    printf "%s\t%s\t%d\n", $kmer, $chrName, $chrStart if($kmer !~ /[^ACGT]/);
    ++$chrStart;
  }
}

##########################################################################

# read in the fasta files
while (my $file = shift) {
  my $sequence = "";
  if ($file =~ m/stdin/) {
    open (FH, "</dev/stdin") or die "can not read stdin";
  } elsif ($file =~ m/.gz$/) {
    open (FH, "gzcat $file|") or die "can not gzcat $file";
  } else {
    open (FH, "<$file") or die "can not read $file";
  }
  while (my $line = <FH>) {
    # skip comment lines
    next if ($line =~ m/^\s*#/);
    # skip empty lines
    next if ($line =~ m/^\s*$/);
    chomp $line;
    if ($line =~ m/^>/) {
      if (length($sequence)) {
        countKmers($sequence);
        $sequence = "";
      }
      chomp $line;
      $chrName = $line;
      $chrName =~ s/^>//;
      $chrStart = 0;
    } else {
      $sequence .= uc($line);
    }
  }
  if (length($sequence)) {
     countKmers($sequence);
     $sequence = "";
  }
  close (FH);
}

File: src/hg/utils/automation/kmerPrint.py
#!/cluster/software/bin/python3

import sys
import gzip

def countKmers(sequence, chrName, kmerSize, chrStart):
    """Count and print kmers from a sequence."""
    sequenceLen = len(sequence)
    for i in range(sequenceLen - kmerSize + 1):
        kmer = sequence[i:i + kmerSize]
        if all(base in "ACGT" for base in kmer):
            print(f"{kmer}\t{chrName}\t{chrStart}")
        chrStart += 1

def processFastaFiles(kmerSize, fastaFiles):
    """Process each FASTA file, reading sequences and generating kmers."""
    chrName = ""
    chrStart = 0
    for file in fastaFiles:
        sequence = ""
        if file == "stdin":
            fileHandle = sys.stdin
        elif file.endswith(".gz"):
            fileHandle = gzip.open(file, "rt")
        else:
            fileHandle = open(file, "r")

        with fileHandle as fh:
            for line in fh:
                line = line.strip()
                if not line or line.startswith("#"):  # Skip empty and comment lines
                    continue
                if line.startswith(">"):
                    if sequence:
                        countKmers(sequence, chrName, kmerSize, chrStart)
                        sequence = ""
                    chrName = line[1:]  # Remove ">"
                    chrStart = 0
                else:
                    sequence += line.upper()
            if sequence:
                countKmers(sequence, chrName, kmerSize, chrStart)
                sequence = ""

def main():
    """Main function to parse arguments and execute the script."""
    if len(sys.argv) < 3:
        sys.stderr.write(
            "usage: kmerPrint.py <N> <fastaFile.fa> [moreFastaFiles.fa] | gzip -c > result.gz\n"
            "where <N> is kmer size\n"
            "followed by fasta file names (can be 'stdin')\n"
        )
        sys.exit(255)

    kmerSize = int(sys.argv[1])
    fastaFiles = sys.argv[2:]

    processFastaFiles(kmerSize, fastaFiles)

if __name__ == "__main__":
    main()
